File: 2021년/9월/0928 소풍.cpp
#include <iostream>
using namespace std;

int c,n,m;
bool isFriends[10][10];
bool couple[10];
int result;

void SOLUTION()
{
	int start_idx = -1;
	for(int i=0;i<n;i++)
	{
		if(!couple[i])
		{
			start_idx = i;
			break;
		}
	}
	if(start_idx == -1)
	{
		result++;
		return;
	}

	for(int i=start_idx+1;i<n;i++)
	{
		if(!couple[i] and isFriends[start_idx][i])
		{
			couple[i] = couple[start_idx] = true;
			SOLUTION();
			couple[i] = couple[start_idx] = false;
		}
	}
}

void INPUT()
{
	scanf("%d", &c);
	while(c--)
	{
		scanf("%d %d", &n, &m);
		
		for(int i=0;i<n;i++)
		{
			couple[i] = false;
			for(int j=0;j<n;j++) isFriends[i][j] = false;
		}
		
		for(int i=0;i<m;i++)
		{
			int x,y;
			scanf("%d %d", &x, &y);
			isFriends[x][y] = true;
			isFriends[y][x] = true;
		}
		
		result = 0;
		SOLUTION();
		printf("%d\n", result);
	}
}

int main()
{
	INPUT();
}

File: 2021년/9월/0928 소풍.py
# 알고스팟_소풍_브루트포스_하

import sys
input = sys.stdin.readline


def dfs():
    global result
    flag = True
    for i in range(n):
        if couple[i] == False:
            start_idx = i
            flag = False
            break
    if flag:
        result += 1
        return

    for i in range(start_idx+1, n):
        if couple[i] == False and isFriend[start_idx][i]:
            couple[i] = couple[start_idx] = True
            dfs()
            couple[i] = couple[start_idx] = False


for _ in range(int(input())):
    n, m = map(int, input().split())
    couple = [False for _ in range(n)]
    isFriend = [[False]*n for _ in range(n)]
    lst = list(map(int, input().split()))
    for i in range(0, len(lst), 2):
        isFriend[lst[i]][lst[i+1]] = True
        isFriend[lst[i+1]][lst[i]] = True

    result = 0
    dfs()
    print(result)
